### PR TITLE
Scope cancellation support to stopping processor

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/21153

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs [hunk 1]
```diff
@@ -298,10 +298,9 @@ protected AmqpConnectionScope()
         /// <param name="receiveMode">The <see cref="ServiceBusReceiveMode"/> used to specify how messages are received. Defaults to PeekLock mode.</param>
         /// <param name="sessionId">The session to connect to.</param>
         /// <param name="isSessionReceiver">Whether or not this is a sessionful receiver.</param>
+        /// <param name="isProcessor">Whether or not the receiver is being created for a processor.</param>
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:301]: We're starting to accumulate a few of these `isBlah` members.   Should we consider some kind of `ClientType` enumeration to make it easier to extend in the future?

**Review by jsquire — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:301]: I removed a lot of the usages of these flags by storing off instance fields on the AmqpReceiver.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs [hunk 2]
```diff
@@ -1045,12 +1052,20 @@ static async Task Open(AmqpObject target, TimeSpan timeout, TaskCompletionSource
...
 
                     default:
                         throw;
                 }
             }
+            finally
+            {
+                if (isProcessor)
+                {
+                    registration.Dispose();
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:1066]: Missing null check?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 3]
```diff
@@ -412,6 +415,13 @@ private static void CloseLink(RequestResponseAmqpLink link)
 
                 throw; // will never be reached
             }
+            finally
+            {
+                if (_isProcessor)
+                {
+                    registration.Dispose();
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:422]: Missing null check?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs (hunk 2 above)
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:1066]: Ah never mind it is a struct. Forgot that

**Pull request merged**